### PR TITLE
test to method mergeSort()

### PR DESCRIPTION
**Pull request by Sofasmile**

### CODE REVIEW

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            {-10, 10, -9, 9, -8, 8, -4, 4, -1, 1},
+            {4, 2, -9, 6, -3, 0, 12, 34, 25, 11},
+    };
+
+    @Test
+    public void mergeSort() {
+        MergeSort sort = new MergeSort();
+        for (int i = 0; i < 5; i++) {
+            int[] actualResult = sort.mergeSort(array[i]);
+            int[] expectedResult = Arrays.stream(array[i]).sorted().toArray();
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: Arrays.sort(array); ?

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package core.basesyntax;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+public class MergeSortTest {
+    private static final int[][] array = {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: testArrays

File: src/main/java/core/basesyntax/MergeSort.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax;
+
+public class MergeSort {
+    /**
+     * Реализуйте метод int[] mergeSort(int[] array), который вернёт отсортированный
+     * массив целых чисел с помощью сортировки слиянием.
+     */
+    public int[] mergeSort(int[] array) {
```
Comment (roman-mate) [inline src/main/java/core/basesyntax/MergeSort.java]: Should we merge it in-place? At least that's how Arrays.sort does it.

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 1 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: Arrays.sort(array) doesn't return array, it changes our global array. and we can't print an array which will not pass the test.

File: src/main/java/core/basesyntax/MergeSort.java (hunk 3 above)
Comment (Sofasmile) [reply src/main/java/core/basesyntax/MergeSort.java]: what do you mean? Sorry, I don't understand) I need create implementation like to the method sort in Arrays class?

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 2 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/main/java/core/basesyntax/MergeSort.java (hunk 3 above)
Comment (roman-mate) [reply src/main/java/core/basesyntax/MergeSort.java]: Arrays class has method sort, which sorts passed in array. Methods which change arguments passed into those methods are kind of anti-patterns but that's how Java Arrays.sort does it. And it this case it feels okay. 

Creating copy of array in most cases won't be needed.

Comment (Sofasmile) [reply src/main/java/core/basesyntax/MergeSort.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    @Test
+    public void mergeSort() {
+        MergeSort sort = new MergeSort();
+        for (int i = 0; i < 5; i++) {
+            sort.mergeSort(testArrays[i]);
+            for (int j = 0; j < testArrays[i].length; j++) {
+                Map<Integer, Integer> expectedMap = new HashMap();
+                for (Integer number : defaultArrays[i]) {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: this loop can be extracted into a helper method.

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MergeSortTest {
+    private static final int[][] defaultArrays = {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: try to get rid of duplicated arrays.

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 4 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 5 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            {-10, 10, -9, 9, -8, 8, -4, 4, -1, 1},
+            {4, 2, -9, 6, -3, 0, 12, 34, 25, 11},
+    };
+
+    private static final int[][] defaultArrays = testArrays.clone();
+
+    @Test
+    public void mergeSort() {
+        MergeSort sort = new MergeSort();
+        for (int i = 0; i < 5; i++) {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]:
```suggestion
        for (int i = 0; i < testArrays.length; i++) {
```

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            sort.mergeSort(testArrays[i]);
+            for (int j = 0; j < testArrays[i].length; j++) {
+                Map<Integer, Integer> expectedMap = transformToHashMap(i);
+                Map<Integer, Integer> actualMap = transformToHashMap(i);
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: are expectedMap and actualMap different?

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    };
+
+    private static final int[][] defaultArrays = testArrays.clone();
+
+    @Test
+    public void mergeSort() {
+        MergeSort sort = new MergeSort();
+        for (int i = 0; i < 5; i++) {
+            sort.mergeSort(testArrays[i]);
+            for (int j = 0; j < testArrays[i].length; j++) {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: is this loop needed?

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 6 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 8 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 7 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            }
+
+            Assert.assertEquals(
+                    "Test failed with elements: " + Arrays.toString(testArrays[i]),
+                    expectedMap,
+                    actualMap);
+        }
+    }
+
+    private Map<Integer, Integer> transformToHashMap(int[] array) {
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: suggestion: countEntries.

File: src/test/java/core/basesyntax/MergeSortTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            {4, 2, -9, 6, -3, 0, 12, 34, 25, 11},
+    };
+
+    private static final int[][] defaultArrays = testArrays.clone();
+
+    @Test
+    public void mergeSort() {
+        MergeSort sort = new MergeSort();
+        for (int i = 0; i < testArrays.length; i++) {
+            sort.mergeSort(testArrays[i]);
```
Comment (roman-mate) [inline src/test/java/core/basesyntax/MergeSortTest.java]: if you compute expectedMap before sorting the array, yoy won't need defaultArrays.

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 9 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

File: src/test/java/core/basesyntax/MergeSortTest.java (hunk 10 above)
Comment (Sofasmile) [reply src/test/java/core/basesyntax/MergeSortTest.java]: done

**Pull request merged**